### PR TITLE
Parallel deployments

### PR DESCRIPTION
**Pull request by mpolden**
FYI @bratseth @tokle

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        if (deploymentSteps.isEmpty()) { // Deployment spec not yet available
+            return false;
+        }
+        DeploymentSpec.Step lastStep = deploymentSteps.get(deploymentSteps.size() - 1);
+        return fromJob(job, application).get().equals(lastStep);
+    }
+
+    /** Returns jobs for given deployment spec, in the order they are declared */
+    public List<JobType> jobsFrom(DeploymentSpec deploymentSpec) {
+        if (deploymentSpec.steps().isEmpty()) {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java]: I think you should remove this if block. DeploymentSpec.steps() should be authorative on which steps to perform.

**Review by bratseth — CHANGES_REQUESTED** [review]
Otherwise LGTM.

Good call to separate ordering out to another class, and very easy to follow the code. 💯

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java (hunk 1 above)
Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java]: Yes, it stems from the fact that a test breaks encapsulation by calling into DeploymentTrigger directly. I'll fix the test and remove the empty check.

**Comment by jonmv** [comment]
Nothing to add.

**Comment by mpolden** [comment]
@bratseth PTAL

**Comment by mpolden** [comment]
Realized that we now need to handle jobs that complete out of order (56a639a996a25dc4c77529017c18d11cb7db41c8).

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java [hunk 2]
```diff
@@ -130,6 +149,68 @@ public void deploymentsSpecWithDelays() {
...
+        // Last declared job completes first
+        tester.notifyJobCompletion(DeploymentJobs.JobType.productionUsWest1, app, true);
+        assertTrue("Change is present as not all jobs are complete",
+                   tester.applications().require(app.id()).deploying().isPresent());
+
+        // All jobs complete
+        tester.notifyJobCompletion(DeploymentJobs.JobType.productionUsEast3, app, true);
+        assertFalse("Change has been deployed",
+                    tester.applications().require(app.id()).deploying().isPresent());
+    }
```
[the remark this answers is not in the document]
Comment (mpolden) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java]: Yes, will fix.

[the remark this answers is not in the document]
Comment (bratseth) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java]: Don't we need to ensure in general that all parallel jobs complete before progressing to the next step?

[the remark this answers is not in the document]
Comment (mpolden) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java]: Yes, will fix.

**Comment by mpolden** [comment]
PTAL.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java [hunk 3]
```diff
@@ -131,7 +131,15 @@ public boolean isDeployableTo(Environment environment, Optional<Change> change)
     public boolean isDeployed(Change change) {
         return status.values().stream()
                 .filter(status -> status.type().isProduction())
-                .allMatch(status -> isSuccessful(status.type(), change));
+                .allMatch(status -> isDeployed(change, status.type()));
+    }
+
+    /** Returns whether job has successfully deployed change */
+    public boolean isDeployed(Change change,  JobType jobType) {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Changing from isSuccessful to isDeployed seems odd to me.
For test steps it is not whether we deployed we are interested in but whether we completed successfully, and that is what we actually check for. 
(And when we separate out production tests as separate steps (which runs after delay), isDeployed becomes completely wrong).

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java [hunk 4]
```diff
@@ -88,11 +93,23 @@ public boolean isLast(JobType job, Application application) {
...
+
+    /** Returns jobs for the given step */
+    private List<JobType> jobsFrom(DeploymentSpec.Step step) {
+        return step.zones().stream()
                 .map(this::toJob)
                 .collect(Collectors.collectingAndThen(Collectors.toList(), Collections::unmodifiableList));
     }
 
+    /** Returns whether given step has deployed completely */
+    private boolean isDeployed(DeploymentSpec.Step step, Application application) {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java]: Same here, should be names "isSuccessful", or maybe even "completedSuccessfully" imho.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bratseth) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTriggerTest.java]: LGTM functionally.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java (hunk 3 above)
Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Changed, thanks.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java (hunk 4 above)
Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentOrder.java]: Changed, thanks.

**Pull request merged**